File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int n = nums1.length, m = nums2.length;
        if(m < n){
            return findMedianSortedArrays(nums2, nums1);
        }
        int i =0 , j = n;
        while( i <= j){
            int partX = i+ (j-i)/2;
            int partY = (m+n+1)/2-partX;
            int partXMax = (partX==0)? Integer.MIN_VALUE: nums1[partX-1];
            int partXMin = (partX==n)?Integer.MAX_VALUE:nums1[partX];
            int partYMax = (partY==0)? Integer.MIN_VALUE: nums2[partY-1];
            int partYMin = (partY==m)?Integer.MAX_VALUE:nums2[partY];
            if( partYMax <= partXMin && partXMax <= partYMin){
                if((m+n)%2==0){
                    return (double)(Math.min(partXMin, partYMin) + Math.max(partXMax, partYMax))/2.0;
                }else{
                    return (double)(Math.max(partXMax, partYMax));
                }
            }else if(partXMin > partYMax){
                j = partX-1;
            }else{
                i= partX+1;
            }
        }
        return -1;
    }
}

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        if(nums1.size() > nums2.size()){
            return findMedianSortedArrays(nums2, nums1);
        }
        int n=nums1.size(),m=nums2.size();
        int l=0, r=nums1.size();
        while(true){
            int partX = (l+r)/2;
            int partY =(n+m+1)/2-partX;

            int maxX = (partX==0)?INT_MIN:nums1[partX-1];
            int maxY = (partY==0)?INT_MIN:nums2[partY-1];
            int minX=(partX==n)?INT_MAX:nums1[partX];
            int minY=(partY==m)?INT_MAX:nums2[partY];

            if(maxX <= minY && maxY <=minX){
                if((n+m)%2==0){
                    return double(max(maxX,maxY)+min(minX,minY))/2;
                }else{
                    return double(max(maxX,maxY));
                }
            }else if(maxX > minY){
                r=partX-1;
            }else{
                l=partX+1;
            }
        }
        return 1.0;
    }
};